mathesar/reflection.py

```diff
@@ -42,53 +44,101 @@ def reflect_databases():
 
 
 # TODO creating a one-off engine is expensive
-def reflect_schemas_from_database(database_name):
-    engine = create_mathesar_engine(database_name)
+def reflect_schemas_from_database(database):
+    engine = create_mathesar_engine(database.name)
     db_schema_oids = {
         schema['oid'] for schema in get_mathesar_schemas_with_oids(engine)
     }
 
-    database = models.Database.current_objects.get(name=database_name)
+    schemas = []
     for oid in db_schema_oids:
-        schema, boolean = models.Schema.current_objects.get_or_create(oid=oid, database=database)
-    for schema in models.Schema.current_objects.all():
+        schema = models.Schema(oid=oid, database=database)
+        schemas.append(schema)
+    models.Schema.current_objects.bulk_create(schemas, ignore_conflicts=True)
+    deleted_schemas = []
+    for schema in models.Schema.current_objects.all().select_related('database'):
         if schema.database.name == database and schema.oid not in db_schema_oids:
-            schema.delete()
+            deleted_schemas.append(schema.id)
+    models.Schema.current_objects.filter(id__in=deleted_schemas).delete()
     engine.dispose()
 
 
-def reflect_tables_from_schema(schema):
+def reflect_tables_from_schemas(schemas):
+    if len(schemas) < 1:
+        return
+    engine = schemas[0]._sa_engine
+    schema_oids = [schema.oid for schema in schemas]
     db_table_oids = {
-        table['oid']
-        for table in get_table_oids_from_schema(schema.oid, schema._sa_engine)
-    }
-    for oid in db_table_oids:
-        models.Table.current_objects.get_or_create(oid=oid, schema=schema)
-    for table in models.Table.current_objects.filter(schema=schema):
-        if table.oid not in db_table_oids:
-            table.delete()
-
-
-def reflect_columns_from_table(table):
-    attnums = {
-        column['attnum']
-        for column in get_column_attnums_from_table(table.oid, table.schema._sa_engine)
+        (table['oid'], table['schema_oid'])
+        for table in get_table_oids_from_schema(schema_oids, engine)
     }
-    for attnum in attnums:
-        column, created = models.Column.current_objects.get_or_create(
-            attnum=attnum,
-            table=table,
-            defaults={'display_options': None})
-        if not created and column.display_options:
+    tables = []
+    for oid, schema_oid in db_table_oids:
+        schema = next(schema for schema in schemas if schema.oid == schema_oid)
+        table = models.Table(oid=oid, schema=schema)
+        tables.append(table)
+    models.Table.current_objects.bulk_create(tables, ignore_conflicts=True)
+    # Calling signals manually because bulk create does not emit any signals
+    models._create_table_settings(models.Table.current_objects.filter(settings__isnull=True))
```

Comment:
`_create_table_settings` is the signal callback, right? And `models.Table.current_objects.filter(settings__isnull=True)` are newly created tables? Would really like some descriptive variable names to immediately know what's what.

Btw, if we change what's listening for table creation signals, this piece of code would become out-of-sync with signal callbacks. Maintainability/debugging nightmare. Would also love a solution for that.

Reply:
> Btw, if we change what's listening for table creation signals, this piece of code would become out-of-sync with signal callbacks. Maintainability/debugging nightmare. Would also love a solution for that.

Its a known issue with Django, one solution I have used in the past dealt with using a custom bulk_create method so that it is not spread out. But lets visit it when when start using bulk_create in more than one place

Reply:
`_create_table_settings`  is usually called by a signal but can also be triggered manually. What name would you suggest for it?

`models.Table.current_objects.filter(settings__isnull=True)` is basically any table whose setting is `null`. Why would you want a descriptive name for it, the statement already says what it is doing

Reply:
> lets visit it when when start using bulk_create in more than one place

My concern is that things will start to break "when when start using bulk_create in more than one place" and it will be hard to know why.

> _create_table_settings is usually called by a signal but can also be triggered manually. What name would you suggest for it?

`_table_creation_callback` would tell me what I need to know.

I'd probably have it in the `models.Table` namespace (either in its own file, which `Table` doesn't have, or a static class method) and have it call `_create_table_settings`.

> models.Table.current_objects.filter(settings__isnull=True) is basically any table whose setting is null.

`created_tables_that_didnt_trigger_signals` would tell me what I need to know here.

> I would like to understand the rationale behind setting a descriptive name for it because the statement already informs what it is doing

The comment gave me valuable hints, but I still need to parse the code.

Reply:
> My concern is that things will start to break "when when start using bulk_create in more than one place" and it will be hard to know why.

The problem is, it is not desirable to group all signals under one common callback function. For example in this case, `Table` has two signals `sync_table_constraints` and `create_table_settings`. Calling `sync_table_constraints` when creating Tables in bulk is not the desired behaviour as we create constraints in a separate call.